chips/lowrisc/spi_host: refactor and rustify.

twilfredo opened the pull request:
### Pull Request Overview

A few cleanup changes to avoid panics, reduced bloat by ~92B according to `cargo bloat`

Before:
```
 0.0%   0.5%     674B         lowrisc lowrisc::spi_host::SpiHost::handle_interrupt
```

After:
```
0.0%   0.4%     582B         lowrisc lowrisc::spi_host::SpiHost::handle_interrupt
```

Edit after a few more fixes:

```
0.0%   0.5%     632B         lowrisc lowrisc::spi_host::SpiHost::handle_interrupt
```

Could probably use a few more improvments...

### Testing Strategy

Running SPI_HOST tests on QEMU and Verilator

### TODO or Help Wanted
 
N/A

### Documentation Updated

- [x] Updated the relevant files in `/docs`, or no updates are required.

### Formatting

- [x] Ran `make prepush`.

hudson-ayers commented on chips/lowrisc/src/spi_host.rs:
```diff
@@ -243,45 +248,78 @@ impl SpiHost {
-                self.spi_transfer_progress();
+                if let Err(err) = self.spi_transfer_progress() {
+                    //Transfer failed, lets clean up
+                    //Clear all pending errors.
```
```suggestion
                    //Clear all pending interrupts.
```

hudson-ayers commented on chips/lowrisc/src/spi_host.rs:
```diff
@@ -243,45 +248,78 @@ impl SpiHost {
+                    //Something went wrong, reset IP and clear buffers
+                    self.reset_spi_ip();
+                    self.reset_internal_state();
+                    self.client.map(|client| match self.tx_buf.take() {
+                        None => (),
+                        Some(tx_buf) => client.read_write_done(
+                            tx_buf,
+                            self.rx_buf.take(),
+                            self.tx_offset.get(),
+                            Err(err),
+                        ),
+                    });
```
we shouldn't call an upcall from within a downcall like this

twilfredo replied:
> we shouldn't call an upcall from within a downcall like this

Ah okay, I'll fix this up! What is the reasoning for this?

hudson-ayers replied:
It is justified here: https://github.com/tock/tock/blob/master/doc/reference/trd2-hil-design.md#rule-1-dont-make-synchronous-callbacks

twilfredo replied:
thanks!

hudson-ayers reviewed (changes requested):
Mostly looks good but one important issue

twilfredo commented:
Changes:

- Fixed up multiple instances of calling upcalls in downcalls, now they occur in the `handle_interrupt()` fn only.
- Added a new enum to detect spihost (specific to this driver) status.

hudson-ayers reviewed (approved):
bors r+
